mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -358,26 +356,48 @@ pub(crate) async fn block_header_utxo_loop<T: UtxoCommonOps>(
         };
 
         // Check if we need to max the number of headers to be stored in storage.
-        if let Some(max_stored_block_headers) = spv_conf.max_stored_block_headers {
-            let headers_in_storage_count = storage.get_total_block_headers_from_storage().await.unwrap_or(0);
-            let current_total = block_registry.len() as u64 + headers_in_storage_count;
-            let max_stored_block_headers = max_stored_block_headers.get();
-
-            if current_total >= max_stored_block_headers {
-                storage
-                    .remove_block_headers_from_storage((current_total - max_stored_block_headers) as i64)
-                    .await
-                    .error_log();
-            }
-        }
+        calculate_and_remove_headers_from_db(storage, spv_conf.max_stored_block_headers, block_registry.len() as u64)
+            .await
+            .error_log();
 
-        // to be removed
-        // println!("max_stored_block_headers {max_stored_block_headers} current_total{current_total}");
         let sleep = args.success_sleep;
         ok_or_continue_after_sleep!(storage.add_block_headers_to_storage(block_registry).await, sleep);
     }
 }
 
+async fn calculate_and_remove_headers_from_db(
+    storage: &BlockHeaderStorage,
+    max_stored_block_headers: Option<NonZeroU64>,
+    block_registry_len: u64,
+) -> Result<(), BlockHeaderStorageError> {
+    if let Some(max_stored_block_headers) = max_stored_block_headers {
+        let headers_in_storage_count = storage.get_total_block_headers_from_storage().await?;
+        let current_total = block_registry_len + headers_in_storage_count;
+        let max_stored_block_headers = max_stored_block_headers.get();
+
+        if current_total >= max_stored_block_headers {
+            let mut headers_to_remove = 0;
+
+            if current_total > max_stored_block_headers {
+                headers_to_remove = current_total - max_stored_block_headers
+            }
+
+            if current_total == max_stored_block_headers {
+                headers_to_remove = max_stored_block_headers - block_registry_len
+            };
+
+            drop_mutability!(headers_to_remove);
+            if headers_to_remove.is_non_zero() {
+                return storage
+                    .remove_block_headers_from_storage(headers_to_remove as i64)
+                    .await;
+            }
+        }
+    };
```

Comment:
I still don't understand why this can't be something simplified to
```rust
if current_total > max_stored_block_headers {
    let headers_to_remove = current_total - max_stored_block_headers;
    return storage
        .remove_block_headers_from_storage(headers_to_remove as i64)
        .await;
}
```
Please explain your reasons for the additional calculations.

Reply:
This will be changed if this comment https://github.com/KomodoPlatform/atomicDEX-API/pull/1585#discussion_r1081548714 is resolved anyways.